### DISCUSSION TITLE
post/linq-stinks

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#6] [Announcements]
# post/linq-stinks

Mark Heath's Development Blog

https://markheath.net/post/linq-stinks

**Comment by atifaziz** [comment]
Good stuff, Mark. For the last section titled **LINQ Stink #5 - Side Effects**, I thought I'd point a couple of ways in which [MoreLINQ] can help since you already called it out a few times (thanks!).

> There are numerous problems with the above code sample, not least the ugly way were trying to call an asynchronous method in the middle of a pipeline, with the `.Result` property which is a recipe for deadlocks (sadly there is no standard 'asynchronous' version of LINQ at the moment, although there are promising signs that we might see that in C# 8).

This is precisely the case for which [`Await`][Await] exists in MoreLINQ (since 2018). Just replace the initial `Select` with `Await` to avoid blocking synchronously, as show below:

```c#
using MoreLinq.Experimental;

var client = new HttpClient();
var urls = new [] { "https://www.alvinashcraft.com/",
                    "http://blog.cwa.me.uk/",
                    "https://codeopinion.com/"}; // imagine this list has 100s of URLs
var regex = @"<a\s+href=(?:""([^""]+)""|'([^']+)').*?>(.*?)</a>";
urls.Await((u, ct) => client.GetStringAsync(u, ct))
    .SelectMany(h => Regex.Matches(h, regex).Cast<Match>())
    .Where(m => m.Value.Contains("markheath.net"))
    .Select(m => m.Value)
    .ToList()
    .ForEach(l => Console.WriteLine(l));
```

Since the signature of `HttpClient.GetStringAsync` is compatible with the delegate/function expected by `Await`, you could even use a method group and so the call could be further simplified to `urls.Await(client.GetStringAsync)`. While the method sits in the experimental namespace, it's been very [carefully designed, implemented and thoroughly reviewed][#205].

By default, `Await` will run the tasks concurrently and stream the results per its description:

> Creates a sequence query that streams the result of each task in the source sequence as it completes asynchronously.

So this even solves the following issue:

> Another issue with this code is that downloading web pages is inherently parallelizable - it would make sense to download more than one at a time, but it is hard to force that behaviour into a LINQ pipeline like this.

What's more, `Await` also supports further configuration to tailor the concurrent execution. The updated example below shows how those options can be configured (commented):

```c#
using MoreLinq.Experimental;

var client = new HttpClient();
var urls = new [] { "https://www.alvinashcraft.com/",
                    "http://blog.cwa.me.uk/",
                    "https://codeopinion.com/"}; // imagine this list has 100s of URLs
var regex = @"<a\s+href=(?:""([^""]+)""|'([^']+)').*?>(.*?)</a>";
urls.Await(client.GetStringAsync)
  // uncomment following for constrained concurrent execution:
  //.MaxConcurrency(4)
  // or uncomment following for sequential execution:
  //.AsSequential()
  // uncomment following for results to be order by input:
  //.AsOrdered()
    .SelectMany(h => Regex.Matches(h, regex).Cast<Match>())
    .Where(m => m.Value.Contains("markheath.net"))
    .Select(m => m.Value)
    .ToList()
    .ForEach(l => Console.WriteLine(l));
```

Finally, MoreLINQ has [`AwaitCompletion`][AwaitCompletion] that can help with the following issue:

> But another big problem here is that downloading web pages is something that is particularly susceptible to transient errors. Suppose my `urls` array contained 100 websites, and they all successfully downloaded except the last one. With the pipeline shown above, no output whatsoever would be written to the console, as one failure will cause the `ToList` call to fail before we even get to the `ForEach`. Yet for the purposes of this code, we probably would like to see the results from the sites we could download, even if some had failed.

And this is how:

```c#
using MoreLinq.Experimental;

var client = new HttpClient();
var urls = new [] { "https://www.alvinashcraft.com/",
                    "http://blog.cwa.me.uk/",
                    "https://codeopinion.com/"}; // imagine this list has 100s of URLs
var regex = @"<a\s+href=(?:""([^""]+)""|'([^']+)').*?>(.*?)</a>";
urls.AwaitCompletion(client.GetStringAsync, (_, t) => t.IsCompletedSuccessfully
                                                    ? t.Result : null)
  // uncomment following for constrained concurrent execution:
  //.MaxConcurrency(4)
  // uncomment following for sequential execution:
  //.AsSequential()
  // uncomment following for results to be order by input:
  //.AsOrdered()
    .Where(h => h is not null) // filter failed ones
    .SelectMany(h => Regex.Matches(h, regex).Cast<Match>())
    .Where(m => m.Value.Contains("markheath.net"))
    .Select(m => m.Value)
    .ToList()
    .ForEach(l => Console.WriteLine(l));
```

In contrast to `Await`, `AwaitCompletion` will allow projecting a result from a task no matter how that task completed. Per its summary:

> Awaits completion of all asynchronous evaluations irrespective of whether they succeed or failed. An additional argument specifies a function that projects the final result given the source item and completed task.

Similar to `Await`, however, `AwaitCompletion` also has the same options for configuring the actual execution of the tasks.

[MoreLINQ]: https://morelinq.github.io/
[Await]: https://morelinq.github.io/3.3/ref/api/html/Overload_MoreLinq_Experimental_ExperimentalEnumerable_Await.htm
[AwaitCompletion]: https://morelinq.github.io/3.3/ref/api/html/M_MoreLinq_Experimental_ExperimentalEnumerable_AwaitCompletion__3.htm
[#205]: https://github.com/morelinq/MoreLINQ/pull/205

**Reply by markheath** [reply]
thanks Atif, I didn't know about  `Await` and `AwaitCompletion`. I'd probably reach for some of the methods in System.Linq.Async these days for situations like this, but always good to know what other options are available

**Reply by atifaziz** [reply]
> I'd probably reach for some of the methods in System.Linq.Async these days for situations like this

Understandable, but have you already come across something along the lines of `Await` and `AwaitCompletion` in **System.Linq.Async**? Wondering if they're redundant and should be eventually retired in MoreLINQ.
